sklearn/neighbors/classification.py

```diff
@@ -364,27 +383,51 @@ def predict(self, X):
 
         weights = _get_weights(neigh_dist, self.weights)
 
-        y_pred = np.empty((n_samples, n_outputs), dtype=classes_[0].dtype)
-        for k, classes_k in enumerate(classes_):
-            pred_labels = np.zeros(len(neigh_ind), dtype=object)
-            pred_labels[:] = [_y[ind, k] for ind in neigh_ind]
-            if weights is None:
-                mode = np.array([stats.mode(pl)[0]
-                                 for pl in pred_labels[inliers]], dtype=np.int)
-            else:
-                mode = np.array([weighted_mode(pl, w)[0]
-                                 for (pl, w)
-                                 in zip(pred_labels[inliers], weights[inliers])],
-                                dtype=np.int)
+        if self.outputs_2d_ == 'sparse':
+            y_pred_sparse_multilabel = []
 
-            mode = mode.ravel()
+            for k, classes_k in enumerate(classes_):
+                pred_labels = np.array([_y[ind, k].toarray()
+                                        for ind in neigh_ind[inliers]],
+                                       dtype=object)
+                y_pred_k = np.zeros(n_samples, dtype=np.int)
+                mode = self._mode(pred_labels, weights, inliers)
+                y_pred_k[inliers] = mode
 
-            y_pred[inliers, k] = classes_k.take(mode)
+                if outliers and self.outlier_label != 0:
+                    y_pred_k[outliers] = self.outlier_label
 
-        if outliers:
-            y_pred[outliers, :] = self.outlier_label
+                y_pred_sparse_multilabel.append(
+                    csc_matrix(y_pred_k.reshape(-1, 1)))
 
-        if not self.outputs_2d_:
-            y_pred = y_pred.ravel()
+            # Old versions of scipy hstack returns COO formatted matrix
+            y_pred = sparse.hstack(y_pred_sparse_multilabel).tocsc()
+
+        else:
+            y_pred = np.empty((n_samples, n_outputs), dtype=classes_[0].dtype)
+            for k, classes_k in enumerate(classes_):
+                pred_labels = np.array([_y[ind, k] for ind in neigh_ind],
+                                       dtype=object)
+                mode = self._mode(pred_labels[inliers], weights, inliers)
+                y_pred[inliers, k] = classes_k.take(mode)
+
+            if outliers:
+                y_pred[outliers, :] = self.outlier_label
+
+            if not self.outputs_2d_:
+                y_pred = y_pred.ravel()
 
         return y_pred
+
+    def _mode(self, pred_labels, weights, inliers):
```

Comment:
move the function above it is used.

Reply:
I would not pass inliers because we take inliers from weights but not from pred_labels. It seems inconsistant.

we should have either `pred_labels[inliers], weight[inliers]` or `pred_inliers, weight_inliers` passed as argument.
I am more in favor of the second because it will be easier with sparse matrices.